File: backend/app/schemas/maintenance.py
from pydantic import BaseModel, model_validator


class MaintenanceSummary(BaseModel):
    database_url: str
    upload_dir: str
    export_dir: str
    project_count: int
    import_batch_count: int
    progress_item_count: int
    report_export_count: int
    backup_command: str


class RuntimeStatus(BaseModel):
    app_version: str
    run_mode: str = "source"
    runtime_mode: str = "source"
    backend_status: str
    database_exists: bool
    database_path: str
    upload_dir: str
    export_dir: str
    backup_dir: str
    project_count: int | str
    import_batch_count: int | str
    progress_item_count: int | str
    report_export_count: int | str
    last_backup_time: str
    backend_started_at: str
    portable_mode: bool | str
    app_root: str
    data_dir: str
    log_dir: str
    frontend_dist_exists: bool
    is_desktop_shell: bool = False
    frontend_served_by_backend: bool = False
    last_diagnose_time: str
    last_diagnose_log_path: str | None = None
    is_release_package: bool = False
    package_version: str | None = None


class AboutRuntimeInfo(BaseModel):
    app_version: str
    runtime_mode: str = "source"
    run_mode: str
    database_path: str
    data_dir: str
    upload_dir: str
    export_dir: str
    backup_dir: str
    core_capabilities: list[str]
    current_limits: list[str]
    quick_actions: list[dict[str, str]]


class CleanupResponse(BaseModel):
    dry_run: bool
    matched_count: int
    cleaned_count: int
    affected_count: int = 0
    matched_ids: list[int]
    message: str
    cleanup_type: str | None = None
    skipped_count: int = 0
    skipped_reasons: list[str] = []
    log_written: bool = False
    details: list[dict[str, str | int | None]] = []

    @model_validator(mode="after")
    def fill_affected_count(self) -> "CleanupResponse":
        self.affected_count = self.matched_count if self.dry_run else self.cleaned_count
        return self


class DataHealthResponse(BaseModel):
    project_count: int | str = 0
    archived_project_count: int | str = 0
    active_project_count: int | str = 0
    import_batch_count: int | str = 0
    frozen_batch_count: int | str = 0
    unfrozen_batch_count: int | str = 0
    unpublished_batch_count: int | str = 0
    published_batch_count: int | str = 0
    progress_item_count: int | str = 0
    warning_record_count: int | str = 0
    rectification_item_count: int | str = 0
    report_export_count: int | str = 0
    orphan_batch_count: int | str = 0
    orphan_item_count: int | str = 0
    missing_file_count: int | str = 0
    database_size: int | str = 0
    upload_dir_size: int | str = 0
    export_dir_size: int | str = 0
    backup_dir_size: int | str = 0
    total_backup_count: int | str = 0
    incomplete_backup_count: int | str = 0
    temp_file_count: int | str = 0
    maintenance_log_count: int | str = 0


class BackupRecord(BaseModel):
    name: str
    backup_time: str
    has_database: bool
    has_uploads: bool
    has_exports: bool
    has_backup_info: bool
    size: int | str = 0
    validation_status: str
    missing_items: list[str] = []
    info_path: str | None = None
    info_content: str | None = None
    backup_path: str


class BackupRestoreRequest(BaseModel):
    confirm_text: str


class BackupRestoreResponse(BaseModel):
    restored: bool
    message: str
    backup_name: str
    pre_restore_backup_name: str
    pre_restore_backup_path: str
    restored_database: bool
    restored_uploads: bool
    restored_exports: bool
    restart_required: bool = True
    log_written: bool = False


class MaintenanceLogRead(BaseModel):
    id: int
    action: str
    target_type: str | None = None
    target_id: int | None = None
    summary: str
    detail: str | None = None
    created_at: str


class CleanupRequest(BaseModel):
    cleanup_type: str
    dry_run: bool = True

File: frontend/src/types/maintenance.ts
export type MaintenanceSummary = {
  database_url: string
  upload_dir: string
  export_dir: string
  project_count: number
  import_batch_count: number
  progress_item_count: number
  report_export_count: number
  backup_command: string
}

export type CleanupResponse = {
  dry_run: boolean
  cleanup_type?: string | null
  matched_count: number
  cleaned_count: number
  affected_count: number
  matched_ids: number[]
  skipped_count: number
  skipped_reasons: string[]
  log_written: boolean
  details: Array<Record<string, string | number | null | undefined>>
  message: string
}

export type RuntimeStatus = {
  app_version: string
  run_mode: string
  runtime_mode: string
  backend_status: string
  database_exists: boolean
  database_path: string
  upload_dir: string
  export_dir: string
  backup_dir: string
  project_count: number | string
  import_batch_count: number | string
  progress_item_count: number | string
  report_export_count: number | string
  last_backup_time: string
  backend_started_at: string
  portable_mode: boolean | string
  app_root: string
  data_dir: string
  log_dir: string
  frontend_dist_exists: boolean
  is_desktop_shell: boolean
  frontend_served_by_backend: boolean
  last_diagnose_time: string
  last_diagnose_log_path?: string | null
  is_release_package: boolean
  package_version?: string | null
}

export type AboutRuntimeInfo = {
  app_version: string
  runtime_mode: string
  run_mode: string
  database_path: string
  data_dir: string
  upload_dir: string
  export_dir: string
  backup_dir: string
  core_capabilities: string[]
  current_limits: string[]
  quick_actions: Array<{ label: string; path: string }>
}

export type DataHealth = {
  project_count: number | string
  archived_project_count: number | string
  active_project_count: number | string
  import_batch_count: number | string
  frozen_batch_count: number | string
  unfrozen_batch_count: number | string
  unpublished_batch_count: number | string
  draft_batch_count: number | string
  imported_unpublished_batch_count: number | string
  parsed_batch_count: number | string
  published_batch_count: number | string
  progress_item_count: number | string
  warning_record_count: number | string
  rectification_item_count: number | string
  report_export_count: number | string
  orphan_batch_count: number | string
  orphan_item_count: number | string
  missing_file_count: number | string
  database_size: number | string
  upload_dir_size: number | string
  export_dir_size: number | string
  backup_dir_size: number | string
  total_backup_count: number | string
  incomplete_backup_count: number | string
  temp_file_count: number | string
  maintenance_log_count: number | string
}

export type BackupRecord = {
  name: string
  backup_time: string
  has_database: boolean
  has_uploads: boolean
  has_exports: boolean
  has_backup_info: boolean
  size: number | string
  validation_status: string
  missing_items: string[]
  info_path?: string | null
  info_content?: string | null
  backup_path: string
}

export type BackupRestoreResponse = {
  restored: boolean
  message: string
  backup_name: string
  pre_restore_backup_name: string
  pre_restore_backup_path: string
  restored_database: boolean
  restored_uploads: boolean
  restored_exports: boolean
  restart_required: boolean
  log_written: boolean
}

export type MaintenanceLog = {
  id: number
  action: string
  target_type?: string | null
  target_id?: number | null
  summary: string
  detail?: string | null
  created_at: string
}

export type MaintenanceAiCallLog = {
  id: number
  project_id?: number | null
  batch_id?: number | null
  mode: string
  model?: string | null
  source: 'ai' | 'rule_fallback'
  success: boolean
  error_message?: string | null
  input_summary_length?: number | null
  output_length?: number | null
  duration_ms: number
  created_at: string
}
